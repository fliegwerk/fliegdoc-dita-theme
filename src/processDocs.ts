const NEWLINE_IN_CODE_PLACEHOLDER = 'REPLACENINCODE' + Date.now();
const BACKTICK_IN_CODE_PLACEHOLDER = 'REPLACEBINCODE' + Date.now();
const MULTIPLY_IN_CODE_PLACEHOLDER = 'REPLACEMINCODE' + Date.now();
const UNDERSCORE_IN_CODE_PLACEHOLDER = 'REPLACEUINCODE' + Date.now();

export function processDocs(raw: string): string {
	return (
		'<p>' +
		raw
			.split('\r')
			.join('')
			.replace(/[<]/g, '&lt;')
			.replace(/[>]/g, '&gt;')
			.replace(
				/```([a-z])*\n([\w\W]+)\n```/,
				(...results) =>
					`<codeblock outputClass="language-${results[1]}">${results[2]
						.split('\n')
						.join(NEWLINE_IN_CODE_PLACEHOLDER)
						.split('`')
						.join(BACKTICK_IN_CODE_PLACEHOLDER)
						.split('*')
						.join(MULTIPLY_IN_CODE_PLACEHOLDER)
						.split('_')
						.join(UNDERSCORE_IN_CODE_PLACEHOLDER)}</codeblock>`
			)
			.split('\n\n')
			.join('</p><p>')
			.replace(/`(.+?)`/g, '<codeph>$1</codeph>')
			.replace(/_(.+?)_/g, '<i>$1</i>')
			.replace(/\*\*(.+?)\*\*/g, '<b>$1</b>')
			.split(NEWLINE_IN_CODE_PLACEHOLDER)
			.join('\n')
			.split(MULTIPLY_IN_CODE_PLACEHOLDER)
			.join('*')
			.split(UNDERSCORE_IN_CODE_PLACEHOLDER)
			.join('_')
			.split(BACKTICK_IN_CODE_PLACEHOLDER)
			.join('`') +
		'</p>'
	);
}
