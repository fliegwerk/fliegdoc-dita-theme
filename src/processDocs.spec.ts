import { processDocs } from './processDocs';

describe('Documentation processor', () => {
	const testCases = [
		`
\`\`\`ts
An example text

With two newlines
\`\`\`
`,
		`
\`\`\`ts
An example text

With two newlines
\`\`\`

And two paragraphs

outside the code
`
	];

	for (let testCase of testCases) {
		it(
			'should correctly process test case #' + testCase.indexOf(testCase),
			() => {
				expect(processDocs(testCase)).toMatchSnapshot();
			}
		);
	}
});
