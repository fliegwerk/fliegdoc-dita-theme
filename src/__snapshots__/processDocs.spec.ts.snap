// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Documentation processor should correctly process test case #0 1`] = `
"<p>
<codeblock outputClass=\\"language-s\\">An example text

With two newlines</codeblock>
</p>"
`;

exports[`Documentation processor should correctly process test case #0 2`] = `
"<p>
<codeblock outputClass=\\"language-s\\">An example text

With two newlines</codeblock></p><p>And two paragraphs</p><p>outside the code
</p>"
`;
